---
title: "Taxonomic Assignment" 
author: "Augustus Pendleton"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: show
    highlight: default
    keep_md: yes
    theme: journal
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
editor_options: 
  chunk_output_type: console
---
<style>
pre code, pre, code {
  white-space: pre !important;
  overflow-x: scroll !important;
  word-break: keep-all !important;
  word-wrap: initial !important;
}
</style>


```{r setup, include=FALSE}
# For width of code chunks and scroll bar 
options(width=250)

knitr::opts_chunk$set(eval = TRUE, 
                      echo = TRUE, 
                      include = TRUE,
                      warning = FALSE,
                      collapse = FALSE,
                      message = FALSE,
                      dpi=200, dev = "png",
                      engine = "R", # Chunks will always have R code, unless noted
                      error = TRUE,
                      fig.path="../figures/02_Taxonomic_Assignment/",  
                      fig.align = "center") 

```


# Purpose

The purpose of this document is to take the feature count (otu) table we generated with dada2 and assign taxonomy to our ASVs. 


# Load packages, source files, and R objects from DADA2_Processing

```{r}
pacman::p_load(tidyverse, phyloseq, Biostrings, dada2, install = FALSE)

knitr::write_bib(file = "data/02_taxonomy_exports/packages.bib")

# Load in the colors and shapes
source("code/R/plotting_aesthetics.R")

# Load in our feature count table
load("data/01_dada2_exports/seqtab_nc_len.RData")

# And load our meta_data
load("data/01_dada2_exports/meta_track.RData")
```

# Assign Taxonomy with TaxAss

```{r, engine = 'bash', engine.opts='-l'}

mkdir -p data/TaxAss_analysis

# First, clone Taxass into Taxass folder

cd data/TaxAss_analysis # Move into Taxass data folder

git clone git@github.com:McMahonLab/TaxAss.git # Clone Taxass Github Repo

cd TaxAss # Move into git repo

cd FreshTrain-files # Move to database zip folders

unzip FreshTrain15Jun2020silva138.zip # Unzip newest version

cd FreshTrain15Jun2020silva138 # Move into it

# Move files into TaxAss_analysis directory
mv -f silva_nr_v138_taxass.fasta ../../../

mv -f silva_nr_v138_taxass.taxonomy ../../../

mv -f FreshTrain15Jun2020silva138.fasta ../../../

mv -f FreshTrain15Jun2020silva138.taxonomy ../../../

# Move back to root of TaxAss git repo
cd ../../

# Move into tax-scripts
cd tax-scripts

# Move everyting into TaxAss_analysis directory
mv -f * ../../

# Move to TaxAss_analysis directory
cd ../../

# Remove TaxAss git repo
rm -r -f TaxAss


# Put dependencies in path
export PATH=/programs/mothur:$PATH
export PATH=/programs/ncbi-blast-2.13.0+/bin:$PATH

cat $PATH
# Run taxass with 50 threads
./RunSteps_quickie.sh ASVs FreshTrain15Jun2020silva138 silva_nr_v138_taxass 98 80 80 50

# Make directory for taxonomy exports if it doesn't exist yet
mkdir -p ../02_taxonomy_exports

# Move our final result file there
mv -f ASVs.98.80.80.taxonomy ../02_taxonomy_exports
mv -f FreshTrain15Jun2020silva138.taxonomy ../02_taxonomy_exports


# Remove everything else besides our ASVs.fasta file (so repo looks exactly like it did when we started)
ls | grep -xv "ASVs.fasta" | xargs rm

# Move back to project root
cd ../../../
```

# Assign taxonomy using Dada/Silva

Here, we will use the silva database version 138. You'll need to download these databases into the 02_taxonomy_exports folder. 

```{r assign-tax}
# The next line took 2 mins to run
taxa <- assignTaxonomy(seqtab_nc_len, "data/02_taxonomy_exports/silva_nr99_v138.1_train_set.fa.gz", multithread=TRUE)

# the next line took 3 minutes 
taxa <- addSpecies(taxa, "data/02_taxonomy_exports/silva_species_assignment_v138.1.fa.gz")

# Inspect the taxonomy 
taxa_print <- taxa # Removing sequence rownames for display only
rownames(taxa_print) <- NULL
head(taxa_print, 20)
```


## 2. Taxonomy Table 

```{r prepare-tax-table}

##### Prepare tax table 
# Add the ASV sequences from the rownames to a column 
new_tax_tab <- taxa %>%
  as.data.frame() %>%
  rownames_to_column(var = "ASVseqs") 
head(new_tax_tab)

# intution check 
stopifnot(new_tax_tab$ASVseqs == colnames(seqtab_nc_len))

asv_headers <- vector(dim(seqtab_nc_len)[2], mode = "character")

# loop through vector and fill it in with ASV names 

for (i in 1:dim(seqtab_nc_len)[2]) {
  asv_headers[i] <- paste(">ASV", i, sep = "_")
}

# intitution check
head(asv_headers, 20)


##### Rename ASVs in table then write out our ASV fasta file! 
#View(seqtab_nochim)
asv_tab <- t(seqtab_nc_len)
#View(asv_tab)

## Rename our asvs! 
row.names(asv_tab) <- sub(">", "", asv_headers)

# Now let's add the ASV names 
rownames(new_tax_tab) <- rownames(asv_tab)
#View(new_tax_tab)

### Final prep of tax table. Add new column with ASV names 
asv_tax <- 
  new_tax_tab %>%
  # add rownames from count table for phyloseq handoff
  mutate(ASV = rownames(asv_tab)) %>%
  # Resort the columns with select
  dplyr::select(Kingdom, Phylum, Class, Order, Family, Genus, Species, ASV, ASVseqs)

# Intution check
stopifnot(asv_tax$ASV == rownames(asv_tax), rownames(asv_tax) == rownames(asv_tab))
```


```{r taxass taxonomy}
taxass_taxa <- read_csv("data/02_taxonomy_exports/ASVs.98.80.80.taxonomy") %>%
  mutate(across(everything(), ~ifelse(str_detect(.x, "unclassified."), NA, .x)), # Convert unclassified to NAs
         across(everything(), ~str_remove_all(.x, "\\(\\d+\\)"))) %>% # Remove the boostrap values after names
  rename_with(.cols = everything(), ~paste0(.x, "_taxass")) # Change column names

cat("Using newest silva, we failed to classify",sum(is.na(asv_tax)), "slots")

cat("Using taxass, we failed to classify",sum(is.na(taxass_taxa)), "slots")

# Here, we figure out which ASVs were assigned from within Taxass
taxass_lineages <- read_tsv("data/02_taxonomy_exports/FreshTrain15Jun2020silva138.taxonomy", col_names = FALSE) %>% # Read in the tax file from the TaxAss custom database
  separate(X2, sep = ";", into = c("kingdom_taxass","phylum_taxass","class_taxass","order_taxass","lineage_taxass","clade_taxass","tribe_taxass","x")) %>% #Do some name cleaning
  mutate(across(everything(), ~ifelse(str_detect(.x, "unnamed."), NA, .x))) %>% # Instead of unclassified, convert to NA (to match our later output)
  inner_join(taxass_taxa) %>% # Innerjoin with our taxonomy file - only ASVs that were classified with the Taxass custom database (assumed to be well curated) stay
  pull(seqID_taxass) # Find those ASV ids.
taxa_joined <- inner_join(asv_tax, taxass_taxa, by = c("ASV" = "seqID_taxass")) %>%
  mutate(from_taxass = ASV %in% taxass_lineages)

# Next, I want to see if when we assign Taxa from TaxAss, do we get better lineage resolution?
taxa_joined %>%
  mutate(Order_Winner = case_when(is.na(Order)&!is.na(order_taxass)~"Taxass",
                                  !is.na(Order)&is.na(order_taxass)~"Silva"),
         Family_Winner = case_when(is.na(Family)&!is.na(lineage_taxass)~"Taxass",
                                  !is.na(Family)&is.na(lineage_taxass)~"Silva"),
         Genus_Winner = case_when( is.na(Genus)&!is.na(clade_taxass)~"Taxass",
                                  !is.na(Genus)&is.na(clade_taxass)~"Silva"),
         Species_Winner = case_when(is.na(Species)&!is.na(tribe_taxass)~"Taxass",
                                   !is.na(Species)&is.na(tribe_taxass)~"Silva")) %>% # Detect whether TaxAss or Silva assigned an ASV to a level that the other couldn't
  group_by(from_taxass) %>%
  summarize(Order_Win_Perc =   sum(str_detect(Order_Winner, "Taxass"), na.rm = T) / sum(!is.na(Order_Winner)),
            Family_Win_Perc = sum(str_detect(Family_Winner, "Taxass"), na.rm = T) / sum(!is.na(Family_Winner)),
            Genus_Win_Perc =  sum(str_detect(Genus_Winner, "Taxass"), na.rm = T) / sum(!is.na(Genus_Winner)),
            Species_Win_Perc = sum(str_detect(Species_Winner, "Taxass"), na.rm = T) / sum(!is.na(Species_Winner))) # Calculate what percentage of time TaxAss "Won" when is used its custom database
```

Generally, if taxa were assigned out of the Freshtrain custom database, we have better assignment to the Order, Family, and Species level. It is slightly worse than Silva at the genus level. Some things, however, we just can't control.

```{r taxass vs silva}
taxa_joined %>% 
  pivot_longer(cols = !c(ASV, ASVseqs, from_taxass), names_to = "Level", values_to = "Classification") %>%
  group_by(Level) %>%
  summarize(nas = sum(is.na(Classification))) %>%
  mutate(Classifier = ifelse(str_detect(Level, "taxass"), "TaxAss", "Silva"),
         Level = str_remove(Level, "_taxass"),
         Level = str_to_title(Level),
         Level = case_match(Level,
                            "Lineage" ~ "Family",
                          "Clade"~ "Genus",
                          "Tribe" ~ "Species",
                          .default = Level),
         Level = factor(Level, levels = c("Kingdom","Phylum","Class","Order","Family","Genus","Species"))) %>%
  ggplot(aes(x = Level, y = nas, color = Classifier, group = Classifier)) + 
  geom_point() + 
  geom_line() +
  theme_classic() + 
  labs(x = "Phylogenetic Level", y = "Number of Unclassified")

taxa_decided <- taxa_joined %>%
  transmute(ASV = ASV,
            Kingdom = ifelse(from_taxass, kingdom_taxass, Kingdom),
            Phylum = ifelse(from_taxass, phylum_taxass, Phylum),
            Class = ifelse(from_taxass, class_taxass, Class),
            Order = ifelse(from_taxass, order_taxass, Order),
            Family = ifelse(from_taxass, lineage_taxass, Family),
            Genus = ifelse(from_taxass, clade_taxass, Genus),
            Species = ifelse(from_taxass, tribe_taxass, Species),
            ASVseqs = ASVseqs)

cat("Prioritizing TaxAss and then the newest silva, we failed to classify",sum(is.na(taxa_decided)), "slots")

rownames(taxa_decided) <- taxa_decided$ASV
```



```{r saving taxonomy files}
# Write the table 
write.table(taxa_joined, "data/02_taxonomy_exports/ASV_both_taxonomy.tsv", sep = "\t", quote = FALSE, col.names = NA)
write.table(taxa_decided, "data/02_taxonomy_exports/ASV_final_taxonomy.tsv", sep = "\t", quote = FALSE, col.names = NA)
```


## Handoff to phyloseq 
```{r phyloseq-handoff}

# Add names to rownames for phyloseq happiness 
rownames(meta_track) <-meta_track$DNA_ID

raw_physeq <- phyloseq(otu_table(asv_tab, taxa_are_rows = TRUE),
                         sample_data(meta_track),
                         tax_table(as.matrix(taxa_decided)))

raw_physeq

save(raw_physeq, file = "data/02_taxonomy_exports/Raw_Physeq.RData")
```

# Session Information 
```{r}
# Reproducibility
devtools::session_info()
```
