---
title: "Phyloseq Preprocessing" 
author: "Augustus Pendleton"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    code_folding: show
    highlight: default
    keep_md: yes
    theme: journal
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(eval = TRUE, 
                      echo = TRUE, 
                      include = TRUE,
                      warning = FALSE,
                      collapse = FALSE,
                      message = FALSE,
                      dpi=200, dev = "png",
                      engine = "R", # Chunks will always have R code, unless noted
                      error = TRUE,
                      fig.path="../figures/03_PhyloSeq_Preprocessing/",  
                      fig.align = "center") 

```
# Goals of our Pre-Processing File

We will work here in this document to remove any unncessary ASVs or samples, such as: 

1. Mitochondria ASVs. 
2. Chloroplast ASVs.  
3. Deal with the controls:  
    a. ASVs found in the controls.  
    b. Control samples themselves.  
4. Mock ASVs/Mock Community.  
5. Samples that have too few reads for reasonable ecological analysis.


## Load Packages and Functions

```{r load-packages-functions}
# Load Packages
pacman::p_load(tidyverse, phyloseq, microViz, Biostrings, dada2, install = FALSE)

knitr::write_bib(file = "data/03_preprocessing_exports/packages.bib")

source("code/R/plotting_aesthetics.R")

```

## Load the data!

```{r load-data}
# Load in the raw_phyloseq data from DADA2_workflow.Rmd
load("data/02_taxonomy_exports/Raw_Physeq.RData")

raw_physeq

load("data/01_dada2_exports/meta_track.RData")
```


# Remove unnecessary taxa

We will need to remove the ASVs that are mitochondria, chloroplast or within the mock community. 

```{r rm-ASVs}
# Make a new phyloseq object without the mitochondria, chloroplast or mock community

noMitoChloroCJMock_physeq <- 
  raw_physeq %>%
  # Remove the mitochondria ASVs.  #NOTE THERE WAS A TYPO HERE IN CLASS! 
  subset_taxa(Family != "Mitochondria" | is.na(Family)) %>%
  # Remove the chloroplast ASVs 
  subset_taxa( Order !="Chloroplast" | is.na(Order)) %>%
  # # remove the mock community and its ASVs 
  # prune_samples(!sample_names(.) %in% c("Zymo_Mock"), .) %>%
  # Now we will remove any ASVs of count 0
  prune_taxa(taxa_sums(.) > 0, .)

# Let's take a look 
noMitoChloroCJMock_physeq

# How many taxa have we removed so far? 
num_ASVs_rm <- ntaxa(raw_physeq) - ntaxa(noMitoChloroCJMock_physeq)
cat("We removed ", num_ASVs_rm, " ASVs")

# Proportion of original data maintained
prop_ASV_rm <- 100* ntaxa(noMitoChloroCJMock_physeq)/ntaxa(raw_physeq)
cat("We retained ", prop_ASV_rm, "% of taxa")
```


## Remove Control samples 

Next, we will need to remove ASVs that seem to be contamination based on our controls

```{r rm-controls, out.width="100%"}
#3. Control samples.  
# Create a vector with the control samples 
control_samples <- c("PCR_Blanks","AP_D61","AP_D62","AP_D153","AP_D154", "Blank1_10471351", "Blank2_10471351")

# Which taxa are in the control samples? 
control_physeq <- 
  noMitoChloroCJMock_physeq %>%
  # Identify control samples
  subset_samples(., (DNA_ID %in% control_samples)) %>%
  # make sure to remove the ASVs that were not in the controls 
  prune_taxa(taxa_sums(.) > 0, .)

# Check out the phyloseq object 
control_physeq

# vector of control ASVs
control_ASVs <- 
  data.frame(control_physeq@tax_table)$ASV

# Plot the abundance of control ASVs across the 4 samples 
control_physeq %>%
  plot_bar(., "ASV", fill = "ASV")

# Plot the abundance of control ASVs across the 4 samples 
control_physeq %>%
  plot_bar(., "ASV", fill = "DNA_ID")

# They are almost ALL in AP_D62

 # Make a phyloseq object with only the control ASVs in it 
 controlASVs_physeq <- 
   noMitoChloroCJMock_physeq %>%
   subset_taxa(., ASV %in% control_ASVs)
 
 # Check it out
 controlASVs_physeq
 
 # Let's take a look at the plot 
  controlASVs_physeq %>%
   ps_mutate(Sample_or_Control = ifelse(DNA_ID%in%control_samples, "Control","Sample")) %>%
   plot_bar(., "Sample_or_Control", fill="Sample_or_Control")+ 
  facet_wrap(~ASV, scales = "free") + 
   labs(title = "All Control ASV Abundance in Samples vs Controls") + 
   theme(legend.position = "none",
         axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
         axis.title = element_text(size = 4),
         axis.text = element_text(size=2),
         strip.text = element_text(size=3),
         axis.ticks = element_blank(),
         axis.line = element_line(linewidth=0.1)) + 
     scale_fill_manual(values = sampletype_colors)
 
control_table <- controlASVs_physeq %>%
  ps_mutate(Sample_or_Control = ifelse(DNA_ID%in%control_samples, "Control","Sample")) %>%
  psmelt() %>% 
  select(ASV, Sample, Abundance, Sample_or_Control, Family, Genus)

control_num <- length(control_ASVs)
sample_num <- nsamples(noMitoChloroCJMock_physeq) - control_num

sam_con_abun <- control_table %>% 
  group_by(ASV, Sample_or_Control) %>%
  summarize(abund = sum(Abundance)) %>%
  mutate(rel_abund = ifelse(Sample_or_Control == "Sample", abund/sample_num, abund/control_num))

# add taxa info
taxa_abund <- left_join(sam_con_abun, as.data.frame(controlASVs_physeq@tax_table), by = "ASV")

ASVs_to_remove <- taxa_abund %>% 
  group_by(ASV) %>%
  slice_max(rel_abund) %>%
  filter(Sample_or_Control == "Control") %>%
  pull(ASV)

# ASVs that we are dropping:
ASVs_to_remove

# add column indicating whether I am keeping it to visualize

## Actually remove the ASVs! 
# Create Phyloseq object that has control samples removed. 
noCONTROLSMitoChloroCJMock_physeq <- 
  noMitoChloroCJMock_physeq %>%
  subset_taxa(., !(ASV %in% ASVs_to_remove))%>%
  # Now remove control samples
  subset_samples(., (!DNA_ID %in% control_samples)) %>%
  # Remove ASVs with 0 counts. 
  prune_taxa(taxa_sums(.) > 0, .) 

# Check out the phyloseq object 
noCONTROLSMitoChloroCJMock_physeq

# After filtering we have removed a total of how many ASVs? 
total_ASVs_rm <- ntaxa(raw_physeq) - ntaxa(noCONTROLSMitoChloroCJMock_physeq)
total_ASVs_rm

```


**Removing ASVs!**  

- There were `r ntaxa(raw_physeq)` ASVs in our raw physeq

- A total of `r total_ASVs_rm` were removed in our filtering steps.

- Now, we have a total of `r ntaxa(noCONTROLSMitoChloroCJMock_physeq)`

# Evaluating reverse complements

```{r}

# Pull out Taxa Table
tax_table <- tax_table(noCONTROLSMitoChloroCJMock_physeq) %>% as.data.frame()

# Grab ASV Sequences and Names
asv_names <- tax_table$ASV
asv_seqs <- tax_table$ASVseqs

asv_rev_comp <- matrix(nrow=length(asv_seqs), ncol = length(asv_seqs)) # Create empty comparison matrix

rownames(asv_rev_comp) <- asv_names
colnames(asv_rev_comp) <- asv_names

asv_dna <- DNAStringSet(asv_seqs) # Convert sequences to Biostrings

asv_rev_dna <- reverseComplement(asv_dna) # Construct reverse complement

for(i in 1:length(asv_seqs)){ # For each asv sequence...
  match_vec <- asv_dna[i] == asv_rev_dna # Compare it to the reverse complement of every other sequence...
  asv_rev_comp[,i] <- match_vec # Write that logical vector as a new column 
}


cat("For",sum(asv_rev_comp) / 2,"ASVs, the reverse complement will need to be removed") # Find how many TRUEs (matches) we have, divide by 2 because each pair occurs twice
```

# Evaluate the Sequencing Depth 
```{r}
# Calcuate the sum of the seqs in each sample
seqSums_df <- 
  noCONTROLSMitoChloroCJMock_physeq %>%
  otu_table() %>%
  # Take the sum of each column of the ASV matrix 
  colSums() %>%
  data.frame() %>%
  rownames_to_column(var = "DNA_ID")
  
# Fix column name in 2nd column 
colnames(seqSums_df)[2] <- "TotalSeqs"

seqSums_df_meta <- seqSums_df %>%
  left_join(meta_track)

# Plot it 
# histogram plot 
seqSums_df_meta %>%
  ggplot(aes(x = TotalSeqs)) + 
  geom_histogram(bins = 50) + 
  labs(title = "Histogram of Raw Sequence Counts per sample")

seqSums_df_meta %>%
  ggplot(aes(x = TotalSeqs, fill = Depth_Class)) + 
  geom_density(alpha = 0.5)


seqSums_df_meta %>%
  ggplot(aes(x = water_extracted, y = TotalSeqs)) + 
  geom_point()

min <- min(seqSums_df$TotalSeqs)

min_sample <- seqSums_df$DNA_ID[seqSums_df$TotalSeqs==min]
```

# Evaluating Accuracy
 
Here, I compare which ASVs are present in our mock community, versus which ones should be:
```{r eval-accuracy}
# Check the mock commmunity 
mock_physeq <- noCONTROLSMitoChloroCJMock_physeq %>%
  subset_samples(.,(DNA_ID=="Zymo_Mock")) %>%
  prune_taxa(taxa_sums(.) > 0, .)

tax_table(mock_physeq) # Look at what taxa are there

otu_matrix <- otu_table(mock_physeq) %>% as.matrix() # Will use this matrix later

cat("DADA2 inferred", ntaxa(mock_physeq), "ASVs present in the Mock Community.")


#Who are they in the mock community? 

mock_seqs <- tax_table(mock_physeq) %>%
  as.data.frame() %>%
  select(ASV, ASVseqs)


#### Compare our ASVs from the mock community to the reference fasta!
mock_reference <- getSequences("data/03_preprocessing_exports/mock_amplicons.fasta")

matches <- mock_seqs %>%
  rowwise() %>%
  mutate(Match = any(grepl(ASVseqs, mock_reference))) # Check each asv to the mock references

match_asv <- matches %>%
  filter(Match) %>%
  pull(ASV) # Pull out asvs that did match reference

nonmatch_asv <- matches %>%
  filter(!Match) %>%
  pull(ASV) # Pull out asvs that didn't match reference

total_mock_reads <- sum(otu_matrix)
matched_reads <- sum(otu_matrix[match_asv])
err_reads <- sum(otu_matrix[nonmatch_asv])



cat("DADA2 inferred ", length(nonmatch_asv), "erroneous ASVs out of the mock community")

cat("Erroneous matches represented ", 100*err_reads/total_mock_reads, "% of total reads in the mock")

```

There were 3 erroneous ASVs inferred, including a fairly abundant Verruco (contamination, maybe) and then a low abundance unassigned (sequencing error, perhaps) and a staphylococcal ASV that may have been contamination from processing

See if the Verruco ASV gets removed

Our minimum number of seqs is `r min(seqSums_df$TotalSeqs)`, which occurs in sample `r min_sample`. Our maximum is `r max(seqSums_df$TotalSeqs)`


# Save preprocessed phyloseq!
```{r save-physeq}
preprocessed_physeq <- noCONTROLSMitoChloroCJMock_physeq %>%
  subset_samples(.,(DNA_ID!="Zymo_Mock"))
save(preprocessed_physeq, file = "data/03_preprocessing_exports/preprocessed_physeq.RData")
```


# Session Information 
```{r}
# Reproducibility
devtools::session_info()
```

